File: typescript/src/resources/media.ts
import type { HttpClient } from "../http"
import type {
  Media,
  MediaListParams,
  MediaListResult,
  UploadMediaParams,
} from "../types"

export class MediaResource {
  constructor(private http: HttpClient) {}

  /** List files in a bucket (paginated). */
  async list(
    bucketSlug: string,
    params?: MediaListParams,
  ): Promise<MediaListResult> {
    return this.http.get<MediaListResult>(
      `/buckets/${encodeURIComponent(bucketSlug)}/media`,
      params as Record<string, unknown> | undefined,
    )
  }

  /** Get a single media record. */
  async get(bucketSlug: string, mediaId: string): Promise<Media> {
    return this.http.get<Media>(
      `/buckets/${encodeURIComponent(bucketSlug)}/media/${encodeURIComponent(mediaId)}`,
    )
  }

  /**
   * Upload a file to a bucket. Works in Node and the browser: `body`
   * is a `Blob` or `File`. Returns the full serialized media record
   * (with `url`, `downloadUrl`, image thumbnails when applicable).
   *
   * @example Browser
   * ```ts
   * const file = input.files[0]
   * const media = await sentroy.media.upload("my-bucket", { body: file })
   * console.log(media.url)
   * ```
   *
   * @example Node 18+
   * ```ts
   * import { openAsBlob } from "node:fs"
   * const blob = await openAsBlob("./photo.jpg")
   * const media = await sentroy.media.upload("my-bucket", {
   *   body: blob,
   *   filename: "photo.jpg",
   * })
   * ```
   */
  async upload(
    bucketSlug: string,
    params: UploadMediaParams,
  ): Promise<Media> {
    const form = new FormData()
    const filename =
      params.filename ||
      (params.body as File & { name?: string }).name ||
      "upload.bin"
    form.append("file", params.body, filename)
    if (params.folder) form.append("folder", params.folder)
    if (params.isPublic !== undefined) {
      form.append("public", params.isPublic ? "true" : "false")
    }
    if (params.alt) form.append("alt", params.alt)
    if (params.caption) form.append("caption", params.caption)
    if (params.tags?.length) form.append("tags", params.tags.join(","))

    return this.http.postForm<Media>(
      `/buckets/${encodeURIComponent(bucketSlug)}/media`,
      form,
    )
  }

  /**
   * Delete a file. Cascades through the CDN: S3 objects (original +
   * thumbnails) are removed, then the Media record. If any S3 delete
   * fails the record is kept so you can retry.
   */
  async delete(bucketSlug: string, mediaId: string): Promise<void> {
    await this.http.del<{ deleted: boolean }>(
      `/buckets/${encodeURIComponent(bucketSlug)}/media/${encodeURIComponent(mediaId)}`,
    )
  }

  /**
   * Download a media file. For private buckets this is the only way to
   * fetch the bytes since the CDN public URL won't serve them; the
   * storage app proxies through after auth.
   *
   * Pass `quality` to request a pre-generated thumbnail variant (e.g.
   * `quality: 500` for 500px wide). Falls back to the original if the
   * variant doesn't exist for that file.
   */
  async download(
    bucketSlug: string,
    mediaId: string,
    opts?: { quality?: number | "original" },
  ): Promise<Blob> {
    const quality = opts?.quality
    const query =
      quality && quality !== "original"
        ? { quality: String(quality) }
        : undefined
    const res = await this.http.fetchRaw(
      `/buckets/${encodeURIComponent(bucketSlug)}/media/${encodeURIComponent(mediaId)}/download`,
      { query },
    )
    if (!res.ok) {
      throw new Error(`Download failed with status ${res.status}`)
    }
    return res.blob()
  }
}

File: python/sentroy/media.py
from __future__ import annotations

import os
import urllib.parse
from typing import BinaryIO, Optional, Union

from sentroy._http import _HttpClient
from sentroy.types import Media, MediaListResult


class MediaResource:
    """Upload, list, download, and delete files inside a bucket."""

    def __init__(self, http: _HttpClient) -> None:
        self._http = http

    def list(
        self,
        bucket_slug: str,
        *,
        type: Optional[str] = None,
        folder: Optional[str] = None,
        limit: Optional[int] = None,
        skip: Optional[int] = None,
    ) -> MediaListResult:
        """List files in a bucket (paginated)."""
        query: dict[str, str] = {}
        if type is not None:
            query["type"] = type
        if folder is not None:
            query["folder"] = folder
        if limit is not None:
            query["limit"] = str(limit)
        if skip is not None:
            query["skip"] = str(skip)

        data = self._http.get(
            f"/buckets/{urllib.parse.quote(bucket_slug, safe='')}/media",
            query=query or None,
        )
        return MediaListResult.from_dict(data or {})

    def get(self, bucket_slug: str, media_id: str) -> Media:
        """Get a single media record."""
        data = self._http.get(
            f"/buckets/{urllib.parse.quote(bucket_slug, safe='')}"
            f"/media/{urllib.parse.quote(media_id, safe='')}"
        )
        return Media.from_dict(data)

    def upload(
        self,
        bucket_slug: str,
        *,
        body: Union[bytes, BinaryIO, str],
        filename: Optional[str] = None,
        content_type: Optional[str] = None,
        folder: Optional[str] = None,
        is_public: Optional[bool] = None,
        alt: Optional[str] = None,
        caption: Optional[str] = None,
        tags: Optional[list[str]] = None,
    ) -> Media:
        """Upload a file to a bucket.

        ``body`` can be raw bytes, a file path (str), or a binary file-like
        object. Returns the full serialized media record (including ``url``
        and ``downloadUrl`` when serving from a public bucket).
        """
        if isinstance(body, str):
            if not os.path.isfile(body):
                raise FileNotFoundError(body)
            if filename is None:
                filename = os.path.basename(body)
            with open(body, "rb") as f:
                content = f.read()
        elif isinstance(body, (bytes, bytearray)):
            content = bytes(body)
        else:
            content = body.read()

        if filename is None:
            filename = "upload.bin"

        fields: dict[str, str] = {}
        if folder is not None:
            fields["folder"] = folder
        if is_public is not None:
            fields["public"] = "true" if is_public else "false"
        if alt is not None:
            fields["alt"] = alt
        if caption is not None:
            fields["caption"] = caption
        if tags:
            fields["tags"] = ",".join(tags)

        data = self._http.post_multipart(
            f"/buckets/{urllib.parse.quote(bucket_slug, safe='')}/media",
            file=(filename, content, content_type),
            fields=fields,
        )
        return Media.from_dict(data)

    def download(
        self,
        bucket_slug: str,
        media_id: str,
        *,
        quality: Optional[Union[int, str]] = None,
    ) -> tuple[bytes, str]:
        """Stream a file's bytes back. Works for both public and private
        buckets — private ones are auth-gated through the storage app.

        ``quality`` requests a pre-generated thumbnail width (e.g. ``500``);
        falls back to the original if the variant wasn't generated. Returns
        ``(bytes, content_type)``.
        """
        query: dict[str, str] = {}
        if quality is not None and quality != "original":
            query["quality"] = str(quality)
        return self._http.fetch_raw(
            f"/buckets/{urllib.parse.quote(bucket_slug, safe='')}"
            f"/media/{urllib.parse.quote(media_id, safe='')}/download",
            query=query or None,
        )

    def delete(self, bucket_slug: str, media_id: str) -> None:
        """Delete a file. Cascades through the CDN: S3 objects (original
        + thumbnails) are removed, then the Media record. If any S3 delete
        fails the record is kept so you can retry."""
        self._http.delete(
            f"/buckets/{urllib.parse.quote(bucket_slug, safe='')}"
            f"/media/{urllib.parse.quote(media_id, safe='')}"
        )
